drivers/adc/ad7616/ad7616.c

```diff
@@ -1039,19 +1052,31 @@ int32_t ad7616_setup(struct ad7616_dev **device,
 	if (init_param->crc) {
 		ret = ad7616_enable_crc(dev);
 		if (ret != 0)
-			return ret;
+			goto cleanup;
 	} else {
 		dev->crc = 0;
 	}
 
 	return ad7616_self_test(dev);
+
+cleanup:
+	ad7616_remove(dev);
```

Comment:
so...please take a look at which APIs are called and which ones require a remove operation.
calling remove in this case (for memory which has not been allocated yet) leads to undefined behavior. 
so, for each API call the according remove function should be called only if the memory was allocated previously.

Reply:
That was the idea at first but then I found out that all the remove function called are doing null checks before freeing the memory (except for `axi_clkgen_remove`, hence the check at line 1075):
- [no_os_pwm_remove](https://github.com/analogdevicesinc/no-OS/blob/0b437cd6b8145324dbac3dd2d59dabe2a67284f3/drivers/api/no_os_pwm.c#L87)
- [no_os_spi_remove](https://github.com/analogdevicesinc/no-OS/blob/0b437cd6b8145324dbac3dd2d59dabe2a67284f3/drivers/api/no_os_spi.c#L127). On this one I think there might be a bug because `desc` is dereferenced *before* the null check. I don't have the full context so I might be wrong also
- [no_os_gpio_remove](https://github.com/analogdevicesinc/no-OS/blob/0b437cd6b8145324dbac3dd2d59dabe2a67284f3/drivers/api/no_os_gpio.c#L112)

I can add null check everywhere in the remove function if you prefer, but I will avoid the goto tag cascade as much as possible in the init function IMAO.

During the preparation of my commit it seems that I forgot to include the switch from `no_os_malloc` to `no_os_calloc` during the device creation. I will add it now if we want to keep only one goto tag

Reply:
using calloc will do the trick, so the remove call can be kept as is.

Reply:
Ok then, perfect.

> On this one I think there might be a bug because desc is dereferenced before the null check. I don't have the full context so I might be wrong also

I still think this should be address later

Reply:
I will take care of the spi remove issue.